source/extensions/http/cache/file_system_http_cache/cache_file_fixed_block.h

```diff
@@ -0,0 +1,189 @@
+   * @param sz The size of the serialized headers, key and metadata.
+   */
+  void setHeadersSize(size_t sz) {
+    contents_.header_size_ = toEndianness<ByteOrder::LittleEndian>(static_cast<uint32_t>(sz));
+  }
+
+  /**
+   * sets the size of the serialized trailers in the header block.
+   * @param sz The size of the serialized trailers.
+   */
+  void setTrailersSize(size_t sz) {
+    contents_.trailer_size_ = toEndianness<ByteOrder::LittleEndian>(static_cast<uint32_t>(sz));
+  }
+
+  /**
+   * sets the size of the serialized body in the header block.
+   * @param sz The size of the body data.
+   */
+  void setBodySize(size_t sz) {
+    contents_.body_size_ = toEndianness<ByteOrder::LittleEndian>(static_cast<uint64_t>(sz));
+  }
+
+  /**
+   * the offset from the start of the file to the start of the serialized headers proto.
+   * @return the offset in bytes.
+   */
+  static off_t offsetToHeaders() { return size(); }
+
+  /**
+   * the offset from the start of the file to the start of the body data.
+   * @return the offset in bytes.
+   */
+  off_t offsetToBody() const { return offsetToHeaders() + headerSize(); }
+
+  /**
+   * the offset from the start of the file to the start of the serialized trailers proto.
+   * @return the offset in bytes.
+   */
+  off_t offsetToTrailers() const { return offsetToBody() + bodySize(); }
+
+  /**
+   * the serialized fixed header chunk.
+   * @return a string view over the entire structure of the header chunk.
+   */
+  absl::string_view stringView() const { return {contents_.raw_, size()}; }
+
+  /**
+   * is this a valid cache file header block for the current code version?
+   * @return True if the block's cache version id and file id match the current version.
+   */
+  bool isValid() const;
+
+private:
+  /**
+   * sets the fileId, a fixed value used to identify that this is a cache file.
+   * This is private so only the unit test can call it, as the file ID should be
+   * a compile-time constant in ordinary use.
+   * @param id The file ID to set.
+   */
+  void setFileId(uint32_t id) { contents_.file_id_ = toEndianness<ByteOrder::LittleEndian>(id); }
+
+  /**
+   * sets the cacheVersionId, a value that should be consistent between versions
+   * of the file cache implementation.
+   * This is private so only the unit test can call it, as the version ID should be
+   * a compile-time constant in ordinary use.
+   * @param id The cache version ID to set.
+   */
+  void setCacheVersionId(uint32_t id) {
+    contents_.cache_version_id_ = toEndianness<ByteOrder::LittleEndian>(id);
+  }
+
+  union {
+    struct {
+      uint32_t file_id_;
+      uint32_t cache_version_id_;
+      uint32_t header_size_;
+      uint32_t trailer_size_;
+      uint64_t body_size_;
+    };
+    char raw_[sizeof(uint32_t) * 4 + sizeof(uint64_t)];
+  } contents_;
+  static_assert(sizeof(contents_) ==
+                    sizeof(contents_.file_id_) + sizeof(contents_.cache_version_id_) +
```

Comment:
this would be much simpler and future-proof if you named the struct.

Reply:
Apparently an anonymous struct can't have a named type. Deanonymizing the struct makes all the code horrible. So I just moved the serialization entirely since you were pushing for that anyway.